### ISSUE TITLE
Legacy nvidia driver 340 fails to compile with recent linux kernel 5.4

### ISSUE THREAD

**Issue by ghost** [NixOS/nixpkgs#83492]
**Describe the bug**
I recently changed the system to ~nixos-18.03~ nixos-20.03 channel. Everything went fine except for my nvidia graphics driver. I have an optimus setup with legacy driver and did not change anything to my working 19.09 configuration. Meanwhile I upgraded the whole system successfully with intel instead of nvidia driver. So the upgrade is working, but nvidia is failing.

During upgrade the compilation of nvidia driver is failing.

**Expected behavior**
The upgrade should not fail.

**Relevant part of output**
```
source root is .
setting SOURCE_DATE_EPOCH to timestamp 1527227501 of file ./nvidia-bug-report.sh
patching sources
applying patch /nix/store/zc1arail4x3b8gcwdrvh8ybiib2inlxw-vm_operations_struct-fault.patch
patching file kernel/uvm/nvidia_uvm_lite.c
configuring
no configure script, doing nothing
building
Building linux driver against kernel: /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev
NVIDIA: calling KBUILD...
make[1]: Entering directory '/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source'
make -C /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/build -f /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/Makefile modules
make[2]: Entering directory '/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/build'
make -f /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/Makefile syncconfig
if [ -f /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/.config -o \
         -d /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/include/config -o \
         -d /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/arch/x86/include/generated ]; then \
        echo >&2 "***"; \
        echo >&2 "*** The source tree is not clean, please run 'make ARCH=x86_64 mrproper'"; \
        echo >&2 "*** in /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source";\
        echo >&2 "***"; \
        false; \
fi
make -f /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/scripts/Makefile.build obj=scripts/basic
ln -fsn /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source source
ln: failed to create symbolic link 'source': Permission denied
make[3]: *** [/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/Makefile:513: outputmakefile] Error 1
make[3]: *** Waiting for unfinished jobs....
  gcc -Wp,-MD,scripts/basic/.fixdep.d -Wall -Wmissing-prototypes -Wstrict-prototypes -O2 -fomit-frame-pointer -std=gnu89       -I ./scripts/basic   -o scripts/basic/fixdep /nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/scripts/basic/fixdep.c   
/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/scripts/basic/fixdep.c:410:1: fatal error: opening dependency file scripts/basic/.fixdep.d: Permission denied
  410 | }
      | ^
compilation terminated.
make[4]: *** [scripts/Makefile.host:107: scripts/basic/fixdep] Error 1
make[3]: *** [/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/Makefile:500: scripts_basic] Error 2
make[2]: *** [/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source/Makefile:677: include/config/auto.conf.cmd] Error 2
make[2]: *** [include/config/auto.conf.cmd] Deleting file 'include/config/tristate.conf'
make[2]: unlink: include/config/tristate.conf: Permission denied
make[2]: Leaving directory '/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/build'
make[1]: *** [Makefile:179: sub-make] Error 2
make[1]: Leaving directory '/nix/store/z1bw23c4bicc1q5kghf5np7gs5hkrk32-linux-5.4.27-dev/lib/modules/5.4.27/source'
NVIDIA: left KBUILD.
nvidia.ko failed to build!
make: *** [Makefile:192: nvidia.ko] Error 1
builder for '/nix/store/s4mqxzkh70r67y3z61alzls7ws4gmhfb-nvidia-x11-340.107-5.4.27.drv' failed with exit code 2
cannot build derivation '/nix/store/62nwxvlhlxhf4hiin6bc88v54ciw79fb-etc.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/dnqm6lxnrrb70g3vkhdy8jd2hhmkjs3q-kernel-modules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/i0kjgw80s62r4k4cixmr71xhbn0wv6a9-linux-5.4.27-modules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/da4mbmlgqy7rk1vja9kj20n502iri81s-system-path.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/nwa2m875znygiyahmdg5v234i1r0j958-nixos-system-laptop-20.03beta786.c6839e5ae8c.drv': 1 dependencies couldn't be built
error: build of '/nix/store/nwa2m875znygiyahmdg5v234i1r0j958-nixos-system-laptop-20.03beta786.c6839e5ae8c.drv' failed
```

**Additional context**
Maybe it is a nvidia - kernel - version confilict. According to [Nvidia](https://nvidia.custhelp.com/app/answers/detail/a_id/3142/), Version 340.108 got support for linux kernel 5.4. But NixOS tries to install version 340.107.

**Metadata**
` - system: `"x86_64-linux"`
 - host os: `Linux 5.4.27, NixOS, 20.03beta786.c6839e5ae8c (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.2`
 - channels(root): `"nixos-20.03beta786.c6839e5ae8c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
`

**Comment by vcunat** [comment]
Just in case, with 4.19 it builds for me, so for now you can configure
```nix
{ # in configuration.nix
  boot.kernelPackages = pkgs.linuxPackages_4_19;
}
```

**Comment by thineye** [comment]
@vcunat I tried it before. It is installing as normal but sadly, if I reboot, the 3d nvidia is failed. `glxinfo` shows me no nvidia device. If I am on stable channel, everything is fine, but as soon as I go to 18.03 channel, its failing, if I am using explicitly 4.19 linux kernel or got nvidia driver 340.108 installed. 

I got successfuly nvidia driver 340.108 installed by using the following in `configuration.nix`, but still not recognized after reboot.

```
    nixpkgs.config = {
      allowUnfree = true;
      packageOverrides = super: let self = super.pkgs; in {
        linuxPackages = super.linuxPackages.extend (linuxSelf: linuxSuper: 
        let
          generic = args: linuxSelf.callPackage (import <nixos/pkgs/os-specific/linux/nvidia-x11/generic.nix> args) { };
        in {
          nvidiaPackages = linuxSuper.nvidiaPackages // {
            legacy_340 = generic {
              version = "340.108";
              sha256_32bit = "1jkwa1phf0x4sgw8pvr9d6krmmr3wkgwyygrxhdazwyr2bbalci0";
              sha256_64bit = "06xp6c0sa7v1b82gf0pq0i5p0vdhmm3v964v0ypw36y0nzqx8wf6";
              settingsSha256 = "1zf0fy9jj6ipm5vk153swpixqm75iricmx7x49pmr97kzyczaxa7";
              persistencedSha256 = "0v225jkiqk9rma6whxs1a4fyr4haa75bvi52ss3vsyn62zzl24na";
              useGLVND = false;

              patches = [ <nixos/pkgs/os-specific/linux/nvidia-x11/vm_operations_struct-fault.patch> ];
            };
          };
        });
      };
    };
```

PS: settingsSha256  and persistencedSha256 was randomly inserted by myself, but installation did not complain.

**Comment by vcunat** [comment]
You're switching to 20.03 and not 18.03, I believe.  If the GPU isn't working well, I'd try to gather logs (journalctl and `/var/log/X.0.log`), but I'm afraid I won't be able to help much with it.  It also sounds independent of the original topic.

**Comment by thineye** [comment]
> You're switching to 20.03 and not 18.03, I believe.

That's correct, I edited my answer.

> It also sounds independent of the original topic.

Not necessarily. If the maintainer would update the nvidia driver to version 430.108 and it is not working during running the operating system, I think, this information is helpful.

But in general, maybe you are right and it is a deeper problem. I try to gather some logs on the working system and non working system and compare them.

**Comment by vcunat** [comment]
430.107 on 4.19 should also work... but I currently don't have a Linux machine with Nvidia.

**Comment by thineye** [comment]
> 430.107 on 4.19 should also work... but I currently don't have a Linux machine with Nvidia.

Well, I tried, but, sadly, it is not working.

`jouranalctl -xe | grep nvidia`

gives me:

```
Apr 04 20:55:23 laptop kernel: nvidia 0000:01:00.0: enabling device (0000 -> 0003)
Apr 04 20:55:23 laptop kernel: nvidia 0000:01:00.0: vgaarb: changed VGA decodes: olddecodes=io+mem,decodes=none:owns=none
Apr 04 20:55:23 laptop kernel: [drm] Initialized nvidia-drm 0.0.0 20150116 for 0000:01:00.0 on minor 1
Apr 04 20:55:08 laptop systemd-modules-load[613]: Failed to find module 'nvidia-uvm'
Apr 04 20:55:12 laptop systemd-modules-load[613]: Inserted module 'nvidia'
Apr 04 20:55:12 laptop systemd-modules-load[613]: Failed to find module 'nvidia_modeset'
Apr 04 20:55:12 laptop systemd-modules-load[613]: Failed to find module 'nvidia_drm'
Apr 04 20:55:12 laptop systemd-udevd[645]: nvidia: Process '/nix/store/9si14qjcz8072c0v42zbkglq08s2cg04-bash-4.4-p23/bin/bash -c 'mknod -m 666 /dev/nvidiactl c $(grep nvidia-frontend /proc/devices | cut -d \  -f 1) 255'' failed with exit code 1.
Apr 04 20:55:57 laptop kernel: caller _nv013519rm+0x57/0xa0 [nvidia] mapping multiple BARs
```

and

`glxinfo`

gives me:

```
name of display: :0
Error: couldn't find RGB GLX visual or fbconfig
```

**Comment by wagnerf42** [comment]
hi,

so I've got the exact same problem.
using 430.108 it compiles and all kernel modules get nicely loaded.

no error message anywhere.

however, if I strace glxinfo I see it opens the ligGL.so.1 from
/nix/store/69hgxhjwa3q7izf4zki8bq54i4n0822n-libGL-1.2.0/lib/libGL.so.1
which is in fact in /nix/store/v52v16f9rj2x42zs0l4940ak0dm4x5x3-libglvnd-1.2.0/lib

I'm no opengl expert but is it ok that we use libglvnd while the nix file has a
              useGLVND = false; ??

I tried setting it to true but it won't compile due to missing json files.
